File: backend/src/EducationalPlatform.Nehzat.Infrastructure/Services/CourierReportService.cs
using EducationalPlatform.Nehzat.Application.DTOs;
using EducationalPlatform.Nehzat.Application.Interfaces;
using EducationalPlatform.Nehzat.Domain.Entities;
using EducationalPlatform.Nehzat.Infrastructure.Data;
using Microsoft.EntityFrameworkCore;

namespace EducationalPlatform.Nehzat.Infrastructure.Services;

public class CourierReportService : ICourierReportService
{
    private readonly AppDbContext _db;

    public CourierReportService(AppDbContext db)
    {
        _db = db;
    }

    public async Task<CourierStatsDto> GetCourierStatsAsync(int userId, int courierUserId, DateTime from, DateTime to)
    {
        var fromDate = from.Date;
        var toDate = to.Date;
        var courier = await _db.Users.FindAsync(courierUserId);
        var courierName = courier is null
            ? $"پیک {courierUserId}"
            : $"{courier.FirstName} {courier.LastName}".Trim();

        var stats = await _db.CourierDailyStats
            .Where(s => s.CourierUserId == courierUserId && s.Date >= fromDate && s.Date <= toDate)
            .OrderBy(s => s.Date)
            .ToListAsync();

        if (stats.Count == 0)
        {
            stats = await BuildStatsFromOrdersAsync(courierUserId, fromDate, toDate);
        }

        var totalDeliveries = stats.Sum(s => s.DeliveredCount);
        var totalDistance = stats.Sum(s => s.DistanceKm);
        var totalEarnings = stats.Sum(s => s.EarningsAmount);

        var daysInPeriod = Math.Max(1, (toDate - fromDate).Days + 1);
        var dailyAverage = stats.Count > 0
            ? totalEarnings / stats.Count
            : totalEarnings / daysInPeriod;

        return new CourierStatsDto
        {
            CourierUserId = courierUserId,
            CourierName = courierName,
            PeriodStart = fromDate,
            PeriodEnd = toDate,
            TotalDeliveries = totalDeliveries,
            TotalDistance = Math.Round(totalDistance, 2),
            TotalEarnings = Math.Round(totalEarnings, 0),
            DailyAverage = Math.Round(dailyAverage, 0),
            DailyBreakdown = stats.Select(s => new CourierDailyStatDto
            {
                Date = s.Date,
                DeliveredCount = s.DeliveredCount,
                DistanceKm = s.DistanceKm,
                EarningsAmount = s.EarningsAmount,
                AverageDeliveryMinutes = s.AverageDeliveryMinutes
            }).ToList()
        };
    }

    public async Task<CourierLeaderboardDto> GetLeaderboardAsync(int userId, DateTime from, DateTime to, int limit = 10)
    {
        var fromDate = from.Date;
        var toDate = to.Date;

        var groups = await _db.CourierDailyStats
            .Where(s => s.Date >= fromDate && s.Date <= toDate)
            .GroupBy(s => s.CourierUserId)
            .Select(g => new CourierAggregate
            {
                CourierUserId = g.Key,
                TotalDeliveries = g.Sum(s => s.DeliveredCount),
                TotalEarnings = g.Sum(s => s.EarningsAmount)
            })
            .OrderByDescending(g => g.TotalEarnings)
            .Take(limit)
            .ToListAsync();

        if (groups.Count == 0)
        {
            groups = await BuildLeaderboardFromOrdersAsync(fromDate, toDate, limit);
        }

        var userIds = groups.Select(g => g.CourierUserId).ToList();
        var users = await _db.Users
            .Where(u => userIds.Contains(u.Id))
            .ToDictionaryAsync(u => u.Id);

        var entries = groups
            .Select((g, i) => new CourierLeaderboardEntryDto
            {
                Rank = i + 1,
                CourierUserId = g.CourierUserId,
                CourierName = users.TryGetValue(g.CourierUserId, out var u)
                    ? $"{u.FirstName} {u.LastName}".Trim()
                    : $"پیک {g.CourierUserId}",
                TotalDeliveries = g.TotalDeliveries,
                TotalEarnings = Math.Round(g.TotalEarnings, 0)
            })
            .ToList();

        return new CourierLeaderboardDto { Entries = entries };
    }

    private async Task<List<CourierDailyStat>> BuildStatsFromOrdersAsync(int courierUserId, DateTime from, DateTime to)
    {
        return await _db.Set<MapOrder>()
            .Where(o => o.AssignedToUserId == courierUserId
                        && o.Status == "delivered"
                        && o.DeliveredAt >= from
                        && o.DeliveredAt <= to.AddDays(1))
            .GroupBy(o => o.DeliveredAt!.Value.Date)
            .Select(g => new CourierDailyStat
            {
                CourierUserId = courierUserId,
                Date = g.Key,
                DeliveredCount = g.Count()
            })
            .OrderBy(s => s.Date)
            .ToListAsync();
    }

    private async Task<List<CourierAggregate>> BuildLeaderboardFromOrdersAsync(DateTime from, DateTime to, int limit)
    {
        return await _db.Set<MapOrder>()
            .Where(o => o.AssignedToUserId != null
                        && o.Status == "delivered"
                        && o.DeliveredAt >= from
                        && o.DeliveredAt <= to.AddDays(1))
            .GroupBy(o => o.AssignedToUserId!.Value)
            .Select(g => new CourierAggregate
            {
                CourierUserId = g.Key,
                TotalDeliveries = g.Count(),
                TotalEarnings = 0m
            })
            .OrderByDescending(g => g.TotalDeliveries)
            .Take(limit)
            .ToListAsync();
    }

    private sealed class CourierAggregate
    {
        public int CourierUserId { get; set; }
        public int TotalDeliveries { get; set; }
        public decimal TotalEarnings { get; set; }
    }
}

File: frontend/src/app/core/services/mock/courier-report.service.ts
import { Injectable } from '@angular/core';
import { Observable, of } from 'rxjs';
import { delay } from 'rxjs/operators';
import type {
  CourierStatsDto,
  CourierLeaderboardDto,
  CourierDailyStatDto,
  CourierLeaderboardEntryDto,
} from '../../models/lesson-planner.models';

interface MockCourier {
  userId: number;
  name: string;
  baseDeliveries: number;
  baseEarnings: number;
}

const COURIERS: MockCourier[] = [
  { userId: 101, name: 'محمد رضایی', baseDeliveries: 6, baseEarnings: 850_000 },
  { userId: 102, name: 'علی احمدی', baseDeliveries: 5, baseEarnings: 720_000 },
  { userId: 103, name: 'فاطمه محمدی', baseDeliveries: 4, baseEarnings: 610_000 },
  { userId: 104, name: 'حسین کریمی', baseDeliveries: 3, baseEarnings: 480_000 },
];

function breakdownFor(courier: MockCourier, from: Date, to: Date): CourierDailyStatDto[] {
  const rows: CourierDailyStatDto[] = [];
  const start = new Date(from);
  const end = new Date(to);
  if (start.getTime() > end.getTime()) {
    return rows;
  }
  const cursor = new Date(start);
  let idx = 0;
  while (cursor.getTime() <= end.getTime() && rows.length < 31) {
    const drift = idx % 2 === 0 ? 1 : 0;
    const deliveries = Math.max(1, courier.baseDeliveries - (idx % 3) + drift);
    rows.push({
      date: cursor.toISOString().slice(0, 10),
      deliveredCount: deliveries,
      distanceKm: Math.round(deliveries * 6.4 * 10) / 10,
      earningsAmount: deliveries * 142_000,
      averageDeliveryMinutes: 28 + (idx % 5) * 4,
    });
    idx++;
    cursor.setDate(cursor.getDate() + 1);
  }
  return rows;
}

@Injectable({ providedIn: 'root' })
export class MockCourierReportService {
  getCourierStats(courierId: number, from: Date, to: Date): Observable<CourierStatsDto> {
    const courier = COURIERS.find((c) => c.userId === courierId) ?? COURIERS[0];
    const daily = breakdownFor(courier, from, to);
    const totalDeliveries = daily.reduce((sum, d) => sum + d.deliveredCount, 0);
    const totalDistance = Math.round(daily.reduce((sum, d) => sum + d.distanceKm, 0) * 10) / 10;
    const totalEarnings = daily.reduce((sum, d) => sum + d.earningsAmount, 0);
    return of({
      courierUserId: courier.userId,
      courierName: courier.name,
      periodStart: from.toISOString(),
      periodEnd: to.toISOString(),
      totalDeliveries,
      totalDistance,
      totalEarnings,
      dailyAverage: daily.length > 0 ? Math.round(totalEarnings / daily.length) : 0,
      dailyBreakdown: daily,
    }).pipe(delay(200));
  }

  getCourierLeaderboard(from: Date, to: Date, limit = 10): Observable<CourierLeaderboardDto> {
    const entries: CourierLeaderboardEntryDto[] = COURIERS
      .map((c, i) => ({
        rank: i + 1,
        courierUserId: c.userId,
        courierName: c.name,
        totalDeliveries: c.baseDeliveries * 20,
        totalEarnings: c.baseEarnings * 20,
      }))
      .slice(0, limit);
    return of({ entries }).pipe(delay(200));
  }
}